BLUE-147 POQo Implementation

Glitch18 opened the pull request:
Implements the POQo algorithm

thantsintoe committed:
selectNeighbors function improvement

Glitch18 committed:
feat: TS port of getCorrespondingNodes and verfifyCorrespondingSender

thantsintoe committed:
initial implementation of factTellCorrespondingNodes and factValidateCorrespondingTellSender functions

thantsintoe committed:
fix bug in filtering the corresponding nodes and validating them. added temp logs (which will be removed later)

thantsintoe committed:
refactor factValidateCorrespondingTellSender()

thantsintoe committed:
proposed a new way for verifyCorrespondingSender function

thantsintoe committed:
set correct value for correspondingGlobalOffset

thantsintoe committed:
verifyCorrespondingSender function update for handle edge case

thantsintoe committed:
implement `factTellCorrespondingNodesFinalData` and `factValidateCorrespondingTellFinalDataSender` using FACT algorithm

Glitch18 committed:
feat: Implement factSendCorrespondingCachedAppData

Glitch18 committed:
feat: implement factValidateCorrespondingCachedAppDataSender

Glitch18 committed:
fix: remove unused variable

Glitch18 committed:
fix: remove not required async tag

Glitch18 committed:
fix: factValidateCorrespondingCachedAppDataSender

Glitch18 committed:
feat: only use factValidation when flag is active

thantsintoe committed:
move neighbor execution node check to outside of factValidateCorrespondingTellSender function

thantsintoe committed:
refactor factTellCorrespondingNodesFinalData for better perf

thantsintoe committed:
code refactor after rebases

Glitch18 committed:
feat: add fact function call

Glitch18 committed:
fix: index calculation in fact send cached app data

Glitch18 committed:
fix: target groups for send cached app data

thantsintoe committed:
fix a bug in calculating the cachedAppData corresponding nodes

Glitch18 committed:
feat: implement configs for poqo

Glitch18 committed:
feat: implement poqo tryProduceReceipt

Glitch18 committed:
feat: implement consensing step for poqo

Glitch18 committed:
feat: voteHash calculation for POQo

Glitch18 committed:
feat: implement POQo send vote loop

Glitch18 committed:
feat: gossip handlers related to poqo

Glitch18 committed:
fix: gossip forwarding for poqo receipt

Glitch18 committed:
feat: implement common origin non burst gossip

Glitch18 committed:
fix: calculations in sendCorrespondingCachedAppData

Glitch18 committed:
fix: poqo handler for sendCorrNodesFinalData

Glitch18 committed:
fix: tryProduceReceipt for POQo

arhamj commented on src/state-manager/TransactionConsensus.ts:
```diff
@@ -1024,6 +1024,220 @@ class TransactionConsenus {
+        _tracker: string,
+        msgSize: number
+      ) => {
+        profilerInstance.scopedProfileSectionStart('poqo-send-vote', false, msgSize)
```
We can add checks to ensure that the receiver is an execution group node and is an elected aggregator.

arhamj commented on src/state-manager/TransactionConsensus.ts:
```diff
@@ -1024,6 +1024,220 @@ class TransactionConsenus {
+        msgSize: number
+      ) => {
+        profilerInstance.scopedProfileSectionStart('poqo-send-receipt', false, msgSize)
+        try{
```
Add catch blocks on all endpoints with counters to avoid unexpected crashes.

arhamj commented on src/state-manager/TransactionConsensus.ts:
```diff
@@ -1024,6 +1024,220 @@ class TransactionConsenus {
+        _respond: unknown,
+        _sender: string,
+      ) => {
+        profilerInstance.scopedProfileSectionStart('poqo-data-and-receipt')
```
Verify FACT sender

arhamj replied:
https://linear.app/shm/issue/GOLD-165

arhamj replied:
This is already handled.

Glitch18 committed:
feat: add log for tracking sharing of cachedAppData

kun6fup4nd4 commented on src/state-manager/TransactionConsensus.ts:
```diff
@@ -1024,6 +1024,220 @@ class TransactionConsenus {
+            // validate corresponding tell sender
+            if (_sender == null) {
+              /* prettier-ignore */ if (logFlags.error) this.mainLogger.error(`poqo-data-and-receipt invalid sender for data: ${data.accountId}, sender: ${_sender}`)
+              continue
```
why not break? _sender is not defined in the loop, should be checked outside.

Glitch18 committed:
feat: add logs to POQo endpoints

Glitch18 committed:
feat: populate receiveAppliedReceipt2 in POQo

Glitch18 committed:
fix: edge case causing missing receipt

Glitch18 committed:
refac: prepone check for null cache data

Glitch18 committed:
feat: POQo switch to requiredVotesPercentage config

Glitch18 committed:
feat: add new conditions for initiating gossip

jairajdev commented on src/state-manager/TransactionConsensus.ts:
```diff
@@ -1024,6 +1024,239 @@ class TransactionConsenus {
+          }
+          const collectedVoteHash = payload as AppliedVoteHash
+          // We can reuse the same function for POQo
+          this.tryAppendVoteHash(queueEntry, collectedVoteHash)
```
We should also check that sign owner of the voteHash is from the executionGroupNodes. inside the `tryAppendVoteHash` fn.
A malicious executionGroup node can create bad voteHash and sign by using different keys, we should check if all these sign owner are in the exectionGroupNodes and reject them if they are not there.

Glitch18 committed:
fix: calculate numVotes in POQo

abdulazeem-tk4vr committed:
GOLD-172 Migrating PoQo endpoints to Binary (#172)

* feat: binary migration of popqo endpints

* PoQo Send Receipt Binary

Binary Migration of PoQo Send Receipt : Handler, Serializer and Deserializer Additions

* feat: Add poqoSendReceiptBinary configuration option

The `poqoSendReceiptBinary` configuration option is added to enable binary migration of the PoQo Send Receipt feature. This commit includes the necessary additions to the handler, serializer, and deserializer for the binary migration.

* refactor: Update internal route for sending receipt in TransactionConsensus.ts

* formatting endpoint position

* feat: Add sending final receipt gossip to binary handler in TransactionConsensus.ts

* refactor: Improve error handling in poqoSendVoteBinary handler

* feat: Fix error handling and improve binary migration in poqo  data and receipt

* refactor: Update poqoSendReceiptReq deserialization logic

Refactor the deserialization logic in the `deserializePoqoSendReceiptReq` function in the `PoqoSendReceiptReq.ts` file. Simplify the code by removing the unnecessary `else` block and handling the `confirmOrChallenge` property more efficiently.

* fix: redundant binary endpoint call

* fix: redundant serde functions and enable bin serializtion

---------

Co-authored-by: Kaung Myat Thu <2962000kaungmyatthu@gmail.com>
Co-authored-by: Atharva <atharvakn@gmail.com>

Commit:
GOLD-164: Signature majority validation for non locally produced receipts (#162)

* Added signature majority validation

* refac: minor changes in verifyAppliedReceipt

* feat: add sig verification for binary endpoints

* feat: signature verification for appending vote hash

* fix: poqo endpoint sig verification

---------

Co-authored-by: Atharva <atharvakn@gmail.com>

jairajdev committed:
Move the receipt sig verification check before processing the accounts data in the poqo-data-and-receipt handler (#177)

* Move the receipt sig verification check before processing the accounts data in the poqo-data-and-receipt

* feat: prepone fact CT sender check

---------

Co-authored-by: Atharva <atharvakn@gmail.com>
